cmd/vsphere-xcopy-volume-populator/internal/populator/vib.go

```diff
@@ -0,0 +1,75 @@
+package populator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kubev2v/forklift/cmd/vsphere-xcopy-volume-populator/internal/vmware"
+	"github.com/vmware/govmomi/object"
+	"k8s.io/klog/v2"
+)
+
+const (
+	vibName     = "vmkfstools-wrapper"
+	vibVersion  = "0.1.0"
+	vibLocation = "/bin/vmkfstools-wrapper.vib"
+)
+
+// ensure vib will fetch the vib version and in case needed will install it
+// on the target ESX
+func ensureVib(client vmware.Client, esx *object.HostSystem, datastore string, desiredVibVersion string) error {
+	version, err := getViBVersion(client, esx)
+	if err != nil {
+		return fmt.Errorf("failed to get the VIB version: %w", err)
+	}
+
+	klog.Infof("current vib version on ESXi: %s %s", esx.Name(), version)
+	if version == desiredVibVersion {
+		return nil
+	}
+
+	vibPath, err := uploadVib(client, esx, datastore)
+	if err != nil {
+		return fmt.Errorf("failed to upload the VIB: %w", err)
+	}
+	klog.Infof("uploaded vib to ESXi %s", esx.Name())
+
+	err = installVib(client, esx, vibPath)
+	if err != nil {
+		return fmt.Errorf("failed to install the VIB: %w", err)
+	}
+	klog.Infof("installed vib on ESXi %s version %s", esx.Name(), vibVersion)
+	return nil
+}
+
+func getViBVersion(client vmware.Client, esx *object.HostSystem) (string, error) {
+	r, err := client.RunEsxCommand(context.Background(), esx, "software", "vib", "get", "-n", vibName)
+	if err != nil {
+		return "", err
+	}
+
+	klog.Infof("reply from get vib %v", r)
+	return r[0].Value("Version"), err
+}
+
+func uploadVib(client vmware.Client, esx *object.HostSystem, datastore string) (string, error) {
+	ds, err := client.GetDatastore(context.Background(), datastore)
+	if err != nil {
+		return "", fmt.Errorf("failed to upload file: %w", err)
+	}
+
+	if err = ds.UploadFile(context.Background(), vibLocation, vibName+".vib", nil); err != nil {
```

Comment:
What if the upload fails due to certs/permissions?

Reply:
Note: We will need to document the needed permissions for this.

Reply:
is the permission to upload to a datastore considered less common?